actors/market/src/state.rs

```diff
@@ -662,20 +663,51 @@ impl State {
                     // pretty fast.
                     // Loading into a HashSet could be an improvement for large collections of deals
                     // in a single sector being removed at one time.
-                    let new_deals = existing_deal_ids
+                    let new_deals: Vec<_> = existing_deal_ids
                         .iter()
                         .filter(|deal_id| !deals_to_remove.contains(*deal_id))
                         .cloned()
                         .collect();
-
-                    sector_deals
-                        .set(sector_number, new_deals)
+                    flush = true;
+
+                    if new_deals.is_empty() {
+                        sector_deals.delete(sector_number).with_context_code(
+                            ExitCode::USR_ILLEGAL_STATE,
+                            || {
+                                format!(
+                                    "failed to delete sector deals for {} {}",
+                                    provider, sector_number
+                                )
+                            },
+                        )?;
+                    } else {
+                        sector_deals.set(sector_number, new_deals).with_context_code(
+                            ExitCode::USR_ILLEGAL_STATE,
+                            || {
+                                format!(
+                                    "failed to set sector deals for {} {}",
+                                    provider, sector_number
+                                )
+                            },
+                        )?;
+                    }
+                }
+            }
+            if flush {
+                if sector_deals.is_empty() {
+                    provider_sectors
+                        .delete(provider)
```

Comment:
Could we push this down into save_provider_sector_deals?

Reply:
would that save us anything? because we're operating across multiple `ActorID`s, we'd have to accumulate a list of deletions to process at the end, but the `delete()` operation doesn't flush state, it's in memory, so we'd end up doing the same amount of work, but with an intermediate accumulation list, wouldn't we?

Reply:
My motivation was de-duplication and correctness. Another call to save_provider_sector_deals is also prefixed with a test whether sector_deals is empty. If the test was inside save_provider_sector_deals in the first place this bug wouldn't have happened. That method should never `set()` with an empty value.

I think the accumulation is just the in-memory provider_sectors state.

Reply:
oh, right, I totally mistook your request as moving it "down" to the save point; but you're right, there's good dedupe potential by moving it into the function